feat(chip-group): Add Chip Group component

macandcheese opened the pull request:
**Related Issue:** [#1933](https://github.com/Esri/calcite-components/issues/1933)

###  ~~Opening as a draft~~ Ready for review :) 

## Summary
This still needs work but want to get a branch up so folks can preview and provide feedback.

This PR adds the proposed `chip-group` component and some associated changes to `chip`.
  - `chip-group` can be used for spacing and keyboard navigation of selectable or non-selectable `chip` components.
  - `chip-group` defaults to `selection-mode:none`, but supports `single`, `single-persist`, and `multiple` selection modes.

To do / looking for feedback:
  - Currently, users need to add `selectable` property to child `chip` components to enable the selection capability. This could potentially be managed by the parent `chip-group`: conditionally applied via an internal property based on `selection-mode`, but that creates a weird situation where `selectable` is not documented but the `selected` property is public, since that will always need to be available to a user. Open to thoughts there.
  
  - Currently, once dismissed, the previous chip is focused. If the first chip is dismissed, focus is moved to the "next first chip" as it takes the place - does this make sense?
  
  - Add more tests as needed.
  - Audit of keyboard navigation.
  - Need an a11y audit from @geospatialem :)
  - Style cleanup / design feedback from @SkyeSeitz @ashetland  :)
  - General pattern feedback :)


https://user-images.githubusercontent.com/4733155/208154126-bdc706de-8c25-4d64-b7e7-988952d96043.mov

alisonailea commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,245 @@
+  //
+  //  Private Methods
+  //
+  //--------------------------------------------------------------------------
```
All these private methods look great but potentially really useful to several components. What about pulling them out into utilities?

macandcheese replied:
💯 agreed - Accordion, Dropdown, Stepper use exact or very similar versions of these as well so making a shared utility makes sense.

macandcheese replied:
💯 PR opened - https://github.com/Esri/calcite-components/pull/6086

alisonailea commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,245 @@
+
+    return (
+      <Host>
+        <div class="container" role={role}>
```
If you are assigning a role here make sure to also allow all potentially required aria attributes and events https://developer.mozilla.org/en-US/docs/web/accessibility/aria/roles/radiogroup_role https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/checkbox_role

macandcheese replied:
@geospatialem could you advise on this?  (at your leisure 🏖️...)

When `chip-group` is `selection-mode="multiple`, and as a result has a role of `checkbox` - 

Should we define `aria-checked` value on the `chip-group` such as:

`mixed` (at least one but less than all child `chips` are selected), 
 `true` (all child `chips` are selected), or 
 `false` (no child `chips` are selected)

?

macandcheese replied:
@geospatialem addressed below, resolving this comment :)

driskull commented:
> Currently, users need to add selectable property to child chip components to enable the selection capability. This could potentially be managed by the parent chip-group: conditionally applied via an internal property based on selection-mode, but that creates a weird situation where selectable is not documented but the selected property is public, since that will always need to be available to a user. Open to thoughts there.

I think the parent could set an internal only selectable prop. Yo me that makes sense.

driskull commented:
> Currently, once dismissed, the previous chip is focused. If the first chip is dismissed, focus is moved to the "next first chip" as it takes the place - does this make sense?

I think that makes sense

driskull reviewed (changes requested):
Looking good! Had some initial feedback 💥👏🎸

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,206 @@
+  /**
+   * emitted when the selected items change
+   */
+  @Event({ cancelable: false }) calciteChipGroupChange: EventEmitter<SelectedItems>;
```
Yep, will add.

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,206 @@
+    this.items = updatedItems;
+  }
+
+  @Listen("calciteInternalChipRegister")
```
Yeah I think we can get rid of these internal registration events, probably in a few other components as well - this was a very old pattern.

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -255,21 +378,61 @@ export class Chip
+      case "single-persist":
+        aria = {
+          "aria-checked": toAriaBoolean(this.selected),
+          "aria-labelledby": this.parent,
```
Good catch - `id` of the parent or if that is not present a generated `guid` (if we need to) ? @geospatialem do we have to set the `id` on parent if not supplied or is that up to implementer as long as it's referenced here with `this.parent.id` ?

macandcheese replied:
@geospatialem added some comments on this below, will resolve this one.

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,206 @@
+  //--------------------------------------------------------------------------
+
+  /** created list of Chip items */
+  private items = [];
```
Did a bit of refactoring to remove the need for this, the internal registration events, element sorting, etc.

geospatialem commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,199 @@
+    "multiple" | "single" | "single-persist" | "none",
+    SelectionMode
+  > = "none";
+
```
We might need a required `label` prop (details below on why):

```ts
  /** Accessible name for the component. */
  @Prop() label!: string;
```

macandcheese replied:
Will add - will each child chip then use `aria-labelledby` to reference the parent label? No need to reference an `id` at all?

In the case a `calcite-chip` will not be wrapped in a `calcite-chip-group` (still valid use case), we won't need to set any of: `role`, `aria-checked`, `aria-labelledby`, correct?

geospatialem replied:
We shouldn't need any additional context, the label provides context when first entering the `chip-group`, and context of the child `chip`s and if it selected, or not selected.

macandcheese replied:
Sorry, I meant when a `calcite-chip` is just placed solo on a page, not within a `calcite-chip-group` (so no selectionMode is assigned) - we just don't need to set any of these aria items, correct?

geospatialem replied:
Some of the basic configurations seem to work well as-is, but some of the more complex ones with icons, icons and close buttons, just images, and when used in cards might need some additional support. There's some context missing in those types in both JAWS and NVDA. Will try to diagnose later on this evening.

geospatialem commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,199 @@
+  //--------------------------------------------------------------------------
+
+  render(): VNode {
+    let role = undefined;
+    switch (this.selectionMode) {
+      case "single":
+      case "single-persist":
+        role = "radiogroup";
+        break;
```
The role works well for `"single"` and `"single-persist"`, maybe for the other two (`"none"` and `"multiple"`) we could use the `"group"` `role`?

Something like:
```ts
    let role = "group";
    switch (this.selectionMode) {
      case "single" || "single-persist":
        role = "radiogroup";
        break;
```

macandcheese replied:
Added :)

geospatialem commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,199 @@
+
+    return (
+      <Host>
+        <div class="container" role={role}>
```
To accommodate the roles for context, we'd just need an `aria-label`, which could be accounted for if we have a required `label` prop. 

Unsure if this is a good standard/consistency with other components, but would work in the case of all `selectionModes`. Something like:

```ts
<div aria-label={this.label} class="container" role={role}>
```

macandcheese replied:
Added :)

geospatialem commented on src/demos/chip-group.html:
```diff
@@ -0,0 +1,506 @@
+      <!-- Neutral (default) -->
+      <div class="parent">
+        <div class="child">None</div>
+        <calcite-chip-group>
```
If the `label` prop is in line with consistency across our component, we would need to update all the `chip-group`'s in the demo to include a `label` prop, such as "None" in this case.

macandcheese replied:
Added :)

geospatialem reviewed (commented):
Looking good @macandcheese! 😎 

Have a few comments with some recommendations, if in line with component consistency, pertaining to the `role` for the `"none"` and `"multiple"` `selectionMode`s.

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,206 @@
+  }
+
+  componentDidLoad(): void {
+    if (!this.sorted) {
```
Sorting shouldn't be needed at all anymore with changes.

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -130,6 +194,11 @@ export class Chip
 
   componentDidLoad(): void {
     setComponentLoaded(this);
+    this.itemPosition = this.getItemPosition();
```
Updated.

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,206 @@
+        focusElementInGroup(this.items, item as HTMLCalciteChipElement, "first");
+      }
+    }
+    this.items = updatedItems;
```
Should be handled with a slot change event now 👍

geospatialem commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,197 @@
+import { Scale, SelectionMode } from "../interfaces";
+
+/**
+ * @slot - A slot for adding one or more `calcite-chip`.
```
nitpick: Do we need to pluralize chips?:

> @slot - A slot for adding one or more `calcite-chip`s.

geospatialem commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,197 @@
+  calciteChipInternalSelectionChange: EventEmitter<HTMLCalciteChipElement>;
+
+  /**
+   * emitted when the selected items change
```
Will this new event be public? If so, WDYT about the following:

> Emits when the component's selection changes.

geospatialem commented on src/components/chip/chip.tsx:
```diff
@@ -81,6 +90,29 @@ export class Chip
+    SelectionMode
+  > = "none";
+
+  /** Is the chip selected  */
```
nitpick:

> When `true`, the component is selected.

geospatialem commented on src/components/chip/chip.tsx:
```diff
@@ -115,13 +145,50 @@ export class Chip
+  // --------------------------------------------------------------------------
+
+  /**
+   * Fires when the close button is clicked.
```
nitpick:

> Fires when the component's close button is selected.

geospatialem commented on src/components/chip/chip.tsx:
```diff
@@ -115,13 +145,50 @@ export class Chip
+  @Event({ cancelable: false }) calciteChipClose: EventEmitter<void>;
+
+  /**
+   * Fires when the selected state of the chip changes due to user interaction.
```
nitpick:

> Fires when the selected state of the component changes.

geospatialem replied:
nitpick:

> Fires when the selected state of the component changes.

geospatialem reviewed (commented):
Some doc considerations for style consistency. 📚

macandcheese commented on src/components/chip/chip.scss:
```diff
@@ -257,6 +299,7 @@ slot[name="image"]::slotted(*) {
   }
 }
+
 :host([closed]) {
```
Yerp 🤙

driskull commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,197 @@
+
+  @Listen("calciteInternalChipKeyEvent")
+  calciteInternalChipKeyEvent(event: CustomEvent): void {
+    if (this.el === event.detail.target.parentElement) {
```
It might be better if it were something like...

```
if(event.composedPath().contains(this.el){
```

So that its not relying on the parent structure.

driskull commented on src/components/chip/chip.e2e.ts:
```diff
@@ -173,4 +183,60 @@ describe("calcite-chip", () => {
   });
 
   it("supports translation", () => t9n("calcite-chip"));
+  /*
```
Can we remove this if its not going to be used?

macandcheese replied:
Yep working on adding these in, will remove any comments.

driskull commented on src/components/chip/chip.tsx:
```diff
@@ -147,38 +214,67 @@ export class Chip
+  keyDownHandler(event: KeyboardEvent): void {
+    if (
+      (event as any).path.includes(this.closeButton) &&
+      (event.key === " " || event.key === "Enter")
```
nitpick: can use the `isActivationKey()` util method from key.ts

driskull commented on src/components/chip/chip.tsx:
```diff
@@ -147,38 +214,67 @@ export class Chip
-    this.closeButton?.focus();
+  @Listen("calciteChipInternalSelectionChange", { target: "body" })
+  internalSelectionChangeListener(event: CustomEvent): void {
+    if (!event.detail.parentNode.contains(this.el)) {
```
nitpick: This could be refactored to not depend on parentNode.

driskull reviewed (approved):
👍

macandcheese commented on src/components/chip-group/chip-group.e2e.ts:
```diff
@@ -0,0 +1,368 @@
+import { accessible, renders, hidden } from "../../tests/commonTests";
+import { CSS as CHIPCSS } from "../chip/resources";
+
+// todo tests
```
Yep, working on more tests for this, will remove notes as added 🧪

jcfranco commented on src/components/chip/chip.tsx:
```diff
@@ -256,21 +369,61 @@ export class Chip
+  }
+
+  render(): VNode {
+    let aria = {};
```
Can you type this?

macandcheese replied:
🖖

jcfranco commented on src/components/chip/chip.tsx:
```diff
@@ -256,21 +369,61 @@ export class Chip
+
+  render(): VNode {
+    let aria = {};
+    switch (this.selectionMode) {
```
Simplification (less nesting): 

```ts
const { selectionMode } = this;
let aria = {};
if (selectionMode === "single" || selectionMode === "single-persist") {
  aria = {
    "aria-checked": toAriaBoolean(this.selected),
    "aria-labelledby": this.parent.label,
    role: "radio"
  }
} else if (selectionMode === "multiple) {
  aria = {
    "aria-checked": toAriaBoolean(this.selected),
    "aria-labelledby": this.parent.label,
    role: "checkbox"
  };
}
```

jcfranco commented on src/components/chip/chip.tsx:
```diff
@@ -115,13 +146,48 @@ export class Chip
+  private containerEl: HTMLDivElement;
+
+  /** the containing chip group element */
+  private parent: HTMLCalciteChipGroupElement;
```
Let's get rid of the comment by renaming the variable to express what it will store.

jcfranco commented on src/components/chip/chip.tsx:
```diff
@@ -98,14 +131,12 @@ export class Chip
+  //--------------------------------------------------------------------------
   //
-  //  Private State/Properties
+  //  Private Properties / State
```
Nitpick: Just noticed we're pretty inconsistent with this. Most files use `Private State/Props`.

jcfranco commented on src/components/chip/chip.tsx:
```diff
@@ -115,13 +146,48 @@ export class Chip
     updateMessages(this, this.effectiveLocale);
   }
 
+  /** determine if there is slotted content for styling purposes */
```
Are these comments necessary? We can see their context in the render method.

macandcheese replied:
Nope, removed.

jcfranco commented on src/components/chip/chip.e2e.ts:
```diff
@@ -25,6 +25,16 @@ describe("calcite-chip", () => {
     expect(eventSpy).toHaveReceivedEvent();
   });
 
+  it("should receive focus when clicked clicked", async () => {
```
Can you also set up the `focusable` helper from `commonTests.ts`? Right now, it only takes care of calling `setFocus`, but I think we could enhance it to cover click and tabbing.

jcfranco commented on src/components/chip/chip.tsx:
```diff
@@ -147,41 +213,71 @@ export class Chip
+    if (event.target !== this.parent) {
+      return;
+    }
+    this.determineActiveItem(event.detail);
```
Not for this PR, but I think we should consider moving this logic to the group. This could apply to other components. I'll create an issue to look into this.

macandcheese replied:
Agreed - I can move that for this PR and clean up any unneeded internal events.

jcfranco reviewed (changes requested):
This is looking great, @macandcheese! 🐿️

I know it's a lot, so LMK if you'd like some additional info or extra help to tackle some of these items.

macandcheese commented on stencil.config.ts:
```diff
@@ -22,7 +22,7 @@ export const create: () => Config = () => ({
     { components: ["calcite-card"] },
     { components: ["calcite-checkbox"] },
-    { components: ["calcite-chip"] },
+    { components: ["calcite-chip", "calcite-chip-group"] },
```
Updated 👍

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,199 @@
+  /**
+   * Emits when the component's selection changes.
+   */
+  @Event({ cancelable: false }) calciteChipGroupSelectChange: EventEmitter<void>;
```
Updated.

macandcheese commented on src/components/chip-group/chip-group.e2e.ts:
```diff
@@ -0,0 +1,382 @@
+
+describe("calcite-chip-group", () => {
+  it("renders", async () =>
+    renders("<calcite-chip-group><calcite-chip></calcite-chip><calcite-chip></calcite-chip></calcite-chip-group>", {
```
Prettier-er now 🪄

macandcheese commented on src/components/chip-group/chip-group.e2e.ts:
```diff
@@ -0,0 +1,382 @@
+
+  it("honors hidden attribute", async () => hidden("calcite-chip-group"));
+
+  it("is accessible in selection mode none (default)", async () => {
```
Updated

macandcheese commented on src/components/chip-group/chip-group.e2e.ts:
```diff
@@ -0,0 +1,382 @@
+  it("honors hidden attribute", async () => hidden("calcite-chip-group"));
+
+  it("is accessible in selection mode none (default)", async () => {
+    accessible(
```
👍

macandcheese commented on src/components/chip-group/chip-group.e2e.ts:
```diff
@@ -0,0 +1,382 @@
+    expect(await page.evaluate(() => document.activeElement.id)).toEqual(chip1.id);
+  });
+
+  it("when chips are selectable, and a chip is focused, using tab will focus the close button", async () => {
```
Interesting - could be good to look into this. May affect Combobox chip dismissal - I wonder if it could mistakenly be hit as users try to exit Combobox?

Good idea though, will leave this comment and add to a follow-up issue after PR is in.

macandcheese commented on src/demos/chip-group.html:
```diff
@@ -0,0 +1,504 @@
+
+      <hr />
+      <div class="child">
+        <h2>Managed with external control - example</h2>
```
Leftover from demo presentations, will remove.

macandcheese commented on src/components/chip/chip.e2e.ts:
```diff
@@ -173,4 +183,59 @@ describe("calcite-chip", () => {
 
   it("supports translation", () => t9n("calcite-chip"));
+
+  it("has `selectable` property programmatically assigned when in a selection-mode='multiple' chip group parent", async () => {
```
I removed the `selectable` prop altogether in an update.

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -115,13 +146,48 @@ export class Chip
+  /**
+   * Fires when the selected state of the component changes.
+   */
+  @Event({ cancelable: false }) calciteChipSelect: EventEmitter<void>;
```
Yep, chips can still be used outside `chip-group`. Also seems to follow pattern from `dropdown-item`, `combobox-item` of emitting their own event in addition to the parent.

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -147,41 +213,71 @@ export class Chip
-    this.closeButton?.focus();
+  @Listen("calciteChipInternalSelectionChange", { target: "body" })
+  internalSelectionChangeListener(event: CustomEvent): void {
+    if (event.target !== this.parent) {
```
Removed this entire event.

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -230,23 +332,34 @@ export class Chip
         aria-label={this.messages.dismissLabel}
         class={CSS.close}
-        onClick={this.closeClickHandler}
-        // eslint-disable-next-line react/jsx-sort-props
```
Yep, was working off an older Chip file, has been restored.

geospatialem commented on src/components/chip/chip.tsx:
```diff
@@ -75,12 +88,30 @@ export class Chip
   @Prop({ reflect: true }) scale: Scale = "m";
 
+  /** Accessible name for the component. */
+  @Prop() label: any;
```
@macandcheese Concur on a required `label`, unless there is a way we could assign the text as a `label`?

macandcheese replied:
Text is just slotted, so it can't be relied on to be present - icon only chip for example.

geospatialem replied:
And yes, thanks for the reminder on the icon only `chip`, concur with the required `label` prop. 👍🏻

macandcheese replied:
As discussed with @jcfranco - `label` will not be required, and `value` will remain required to avoid breaking changes and the need for translations internally for components that render `chip` components of their own - Alert, Combobox, Rating. 

These being required or not is pretty inconsistent across all components, so any patterns we end up moving towards can be applied to Chip at that time.

geospatialem replied:
Is the bigger concern around breaking changes for the requirement? We'll definitely need to doc the importance of the `label` prop if its not required, in particular for situations where there's only an icon present.

macandcheese replied:
cc @jcfranco 🤔

jcfranco replied:
> Is the bigger concern around breaking changes for the requirement? We'll definitely need to doc the importance of the label prop if its not required, in particular for situations where there's only an icon present.

Correct. Making it required will cause TS errors to start showing up if it wasn't already being provided for existing chips. 

> These being required or not is pretty inconsistent across all components, so any patterns we end up moving towards can be applied to Chip at that time.

💯, we can start tracking these and batch update them for whenever we have a future breaking release in the works.

macandcheese commented:
@ashetland @SkyeSeitz the design for this should be ready for review after the recent commit - could you give this a look and see if all the combinations of icon / avatar / text / closable / selection look ok?

SkyeSeitz commented:
> @ashetland @SkyeSeitz the design for this should be ready for review after the recent commit - could you give this a look and see if all the combinations of icon / avatar / text / closable / selection look ok?

On it. Looking good thus far- Let's connect Monday to go over padding and spacing 🚀

macandcheese commented on src/components/chip-group/chip-group.e2e.ts:
```diff
@@ -0,0 +1,382 @@
+    expect(eventSpyChip2).toHaveReceivedEventTimes(0);
+    expect(chip1).toHaveAttribute("selected");
+    expect(chip2).not.toHaveAttribute("selected");
+    expect(await element.getProperty("selectedItems")).toHaveLength(1);
```
Updated to use one of the existing patterns, and added an issue to consolidate those in use as discussed: https://github.com/Esri/calcite-components/issues/6713

macandcheese commented:
@jcfranco could you take a look, I think I addressed all the feedback. 

I still need to make some changes to Combobox as it had some custom logic for what is now supplied by Chip Group.

jcfranco commented on src/components/chip/chip.tsx:
```diff
@@ -147,41 +213,71 @@ export class Chip
-  @Method()
-  async setFocus(): Promise<void> {
-    await componentLoaded(this);
+  @Listen("keydown", { capture: true })
```
Is `capture` still applicable here?

macandcheese replied:
Removed

jcfranco reviewed (approved):
Aside from the Fixme comments, this looks good to merge!

Amazing stuff, @macandcheese! 

🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪
🍪🍪🔵🔵🔵🍪🔵🍪🍪🔵🍪🔵🔵🔵🍪🔵🔵🔵🍪🍪🍪🔵🔵🔵🍪🔵🍪
🍪🔵🍪🍪🍪🍪🔵🍪🍪🔵🍪🍪🔵🍪🍪🔵🍪🍪🔵🍪🔵🍪🍪🍪🍪🔵🍪
🍪🔵🍪🍪🍪🍪🔵🔵🔵🔵🍪🍪🔵🍪🍪🔵🔵🔵🍪🍪🍪🔵🔵🍪🍪🔵🍪
🍪🔵🍪🍪🍪🍪🔵🍪🍪🔵🍪🍪🔵🍪🍪🔵🍪🍪🍪🍪🍪🍪🍪🔵🍪🍪🍪
🍪🍪🔵🔵🔵🍪🔵🍪🍪🔵🍪🔵🔵🔵🍪🔵🍪🍪🍪🍪🔵🔵🔵🍪🍪🔵🍪
🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪🍪

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,202 @@
+  //  Private Properties
+  //
+  //--------------------------------------------------------------------------
+  mutationObserver = createObserver("mutation", () => this.updateItems);
```
👀 😄   https://github.com/Esri/calcite-components/pull/6075/files/930a476816baefedabf636b182df7b4cab9d102c#r1137892003

macandcheese commented on src/components/chip-group/chip-group.e2e.ts:
```diff
@@ -0,0 +1,502 @@
+ * @param root0
+ * @param root0.expectedItemIds
+ */
+async function assertSelectedItems(page: E2EPage, { expectedItemIds }: SelectedItemsAssertionOptions): Promise<void> {
```
Opened this issue for this - few other similar testing patterns across components: https://github.com/Esri/calcite-components/issues/6713

macandcheese commented on src/components/chip-group/chip-group.stories.ts:
```diff
@@ -0,0 +1,38 @@
+`;
+
+export const darkThemeRTL_TestOnly = (): string => html`
+  <div style="background-color:#2b2b2b;" dir="rtl">
```
Nope - removed

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -232,48 +304,91 @@ export class Chip
+    return (
+      <Host>
+        <div
+          {...aria}
```
Yep, removed and inlined.

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -157,33 +246,30 @@ export class Chip
 
-  closeClickHandler = (event: MouseEvent): void => {
-    event.preventDefault();
+  private closeHandler = (): void => {
```
Updated

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -232,48 +304,91 @@ export class Chip
-        ref={(el) => (this.closeButton = el)}
+        onClick={this.closeHandler}
+        onKeyDown={this.closeButtonKeyDownHandler}
+        tabIndex={!this.disabled ? 0 : -1}
```
Correct, it can't be reached via keyboard or mouse within the disabled chip. Should this be applied anyway here?

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -75,12 +90,30 @@ export class Chip
   @Prop({ reflect: true }) scale: Scale = "m";
 
+  /** Accessible name for the component. */
+  @Prop() label: any;
```
Finally a chance for me to bump this 2 year old issue: https://github.com/Esri/calcite-components/issues/3267

macandcheese commented on src/components/chip/chip.tsx:
```diff
@@ -117,13 +148,45 @@ export class Chip
     updateMessages(this, this.effectiveLocale);
   }
 
+  @State() private hasContent = false;
```
Correct, this is similar to Button implementation.

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,239 @@
+  @Method()
+  async setFocus(): Promise<void> {
+    await componentLoaded(this);
+    if (!this.disabled && this.selectedItems.length > 0) {
```
👨‍🍳

macandcheese commented on src/components/chip/chip.scss:
```diff
@@ -2,30 +2,49 @@
-  .image-container {
-    @apply h-5 w-5;
-    padding-inline-start: theme("spacing[0.5]");
+  --calcite-chip-spacing-unit-l-internal: 0.25rem;
```
Whoops, reverted.

macandcheese commented on src/components/chip-group/chip-group.tsx:
```diff
@@ -0,0 +1,202 @@
+
+    this.selectedItems = this.items.filter((el) => el.selected);
+
+    if (event) {
```
To clarify, just a boolean that determines whether or not to emit here, correct?

macandcheese commented:
Going to try and merge this in pending a Chromatic review from @ashetland / @SkyeSeitz - there are a few px differences in chip / chip group that need approval. If needed we can continue to adjust small styles but I'd like to get this component in so we can begin building documentation and samples. 

cc @jcfranco is this good to go following Chromatic approval? Anything we need follow up issue for (using `delete` key to facilitate dismissal)?
